### ISSUE TITLE
Cookies being sent to wrong site

### ISSUE THREAD

**Issue by anjackson** [internetarchive/heritrix3#259]
We've now see a few cases where some cookies are being sent over-and-over in requests to lots of different sites. For an example WARC from our domain crawl, we see 

```
WARC/1.0^M
WARC-Type: request^M
WARC-Target-URI: http://gamstop.co.uk/^M
WARC-Date: 2019-06-26T21:57:15Z^M
WARC-Concurrent-To: <urn:uuid:c8c3061e-74aa-4836-8e95-11270677cac7>^M
WARC-Record-ID: <urn:uuid:d51c9542-a658-4901-9445-e5a8e06d4cf3>^M
Content-Type: application/http; msgtype=request^M
Content-Length: 1750^M
^M
GET / HTTP/1.0^M
Connection: Close^M
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8^M
Host: gamstop.co.uk^M
User-Agent: bl.uk_ldfc_bot/3.4.0-20190418 (+http://www.bl.uk/aboutus/legaldeposit/websites/websites/faqswebmaster/index.html)^M
Cookie: Country=GB; EPRAT=1966168438-1561106086442; ESTN=2; PHPSESSID=p4bsl9dkml874kbnrs28isqrl2; PM_unread_Onlineshop="6669,9502,6679,5575,6055,6936,9117,5574,6670,6673,6627,70170,6814,6595,6464,6003,5582,5338,6938,9488,6811,9360,6597,9366,5586,9823,9454,9377,9453,7368,6573,6121,9800,9847,5558,9801,6267,6592,9057,5583,9843,6463,9542,9811,9814,70038,5584,9812,6672,9420,9362,9363,6596,9424,5560,6590,9816,9844,9464,6389,9275,5559,6598,9806,9805,9842,70219,6939,9815,6410,9472,6215,9807,9808,6594,6591,9830,6476,9421,9423,70013,9404,9361,5131,6584,6593,6599,6426,6354,6279,6578,9822,9833,5398,9438,70107,9435,9491,6586,6581,9827,9852,5364,9832,9846,9277,9276,70175,9819,6158,70041,70124,9448,70043,9279,5653,9442,6588,5399,9834,70121,70120,9375,6589,70106,70025,9818,6495,6214,6576,9425,70040,9456,9474,70026,70044,9813,9447,6574,6922,9851,9379,5657,6577,9356,9804,6575,70042,9383,9378,9492,6506,7700,9321,9426,6587,70123,9826,7892,9825,9439,9440,9355,9352,70045,9353,9350,70029,70032,70030,9446,9380,9381,6585,6583,9449,9441,9437,9359,70031,70027,9357,9365,9809,9418,70119,9810,70037,9457,9817,9824,70122,9376,70028,9358,9487,9322,70118,5626,9462,9455,9278,9417,9465,9463,9468,9419,9354,9415,9820,9490,9390,9406,9428,9416,9351,70039,5650,9466,9850,9055,70012,9467,9401,9803,9427,9831,9373,9405,9845,70172,9436,9489,9853,9059,9403,9471,9384,9543,70173,6270,9469,9340,9345,9364,5130,9821,9343,9342,9374,9473,9341,5656,9422,70171,9382,9144,9389,70174,6427,9344,9402,9802,9470"; PSACountry=GB^M
^M
^M
^M
```

This appears to be in large numbers of requests, and because it's large, sometimes causes problems (we got 403 blocked by http://gamstop.co.uk/ in this example.)

We've visited the site before in this crawl, but these cookies didn't turn up there.

**Comment by anjackson** [comment]
Having grepped about in the BDB Cookie Store, plus a bit of Googling, I've found an example that's causing the problem: `PM_unread_Onlineshop`.  It's used on https://www.playmobil.co.uk/, and if you look you get this:

```
# curl -I "https://www.playmobil.co.uk/"
HTTP/1.1 200 OK
Date: Thu, 27 Jun 2019 14:10:56 GMT
Content-Type: text/html;charset=UTF-8
Connection: keep-alive
Set-Cookie: __cfduid=d5f95e31b34b7481bb583bb9882b0532e1561644656; expires=Fri, 26-Jun-20 14:10:56 GMT; path=/; domain=.playmobil.co.uk; HttpOnly
Set-Cookie: dwac_df5bAiaagI3poaaadtZ86YcFHU=CSbfnIfqXZCLdz5WUQKkxjy7Z4pHj6ig6zU%3D|demandaabegb|||GBP|false|Europe%2FLondon|true; Path=/
Set-Cookie: cqcid=cdRzkECta1laPzaFq75To2bmsN; Path=/
Set-Cookie: sid=CSbfnIfqXZCLdz5WUQKkxjy7Z4pHj6ig6zU; Path=/
Set-Cookie: dwsecuretoken_8af1144abc5a7219343b18330f0adea9=g2dOqHMaOL_hfLH0UVmnHc61xRHroJj9vQ==; Version=1; Comment="Demandware Secure Token for site Sites-GB-Site"; Path=/; Secure; HttpOnly
Set-Cookie: dwpersonalization_8af1144abc5a7219343b18330f0adea9=""; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Path=/
Set-Cookie: dwanonymous_8af1144abc5a7219343b18330f0adea9=cdRzkECta1laPzaFq75To2bmsN; Version=1; Comment="Demandware anonymous cookie for site Sites-GB-Site"; Max-Age=15552000; Expires=Tue, 24-Dec-2019 14:10:56 GMT; Path=/
x-dw-request-base-id: GGZMbITHFF0BAAB_
Set-Cookie: __cq_dnt=0; Path=/
Set-Cookie: dw_dnt=0; Path=/
Set-Cookie: PM_unread_Onlineshop="6669,9502,6679,6055,5575,6936,9117,6670,6003,5574,6938,6673,6814,5582,9488,5558,6627,70170,9453,6595,9360,9805,5338,6597,6939,9420,9366,9277,9275,6811,6121,6592,6573,9800,9814,9823,5583,9454,6464,9057,9801,6267,70038,7368,9812,5584,9830,6672,9362,9363,9361,6527,5560,9542,70039,6590,9843,9844,5653,7892,9833,9424,9470,9466,6596,9819,9815,6574,6410,9279,9375,9472,6584,6506,6463,9853,9377,9489,6354,9834,70120,6576,6476,9842,9423,9404,5131,6591,6599,9448,9474,6578,9491,9822,9435,9803,9802,6389,6586,6598,9827,5364,70219,9144,70026,6577,9832,9471,9276,70119,70175,9426,6158,5586,9442,5559,9811,70041,9447,9425,9816,9807,9808,70121,6426,70123,9826,9356,70040,9406,9438,70043,9813,70107,70124,70025,70044,9352,9820,6922,9851,9439,70029,6588,6594,9852,9355,70030,6589,70042,6575,9492,9446,7700,5399,6270,9457,9322,9321,9818,9390,9806,9824,9825,6587,9351,9350,9353,70032,9449,9381,9357,9440,9441,9437,70031,70027,6585,6583,6593,6581,9464,5398,70045,9365,9809,9804,9810,9456,6215,9418,9817,6495,9421,9376,9847,70037,9487,5626,70106,9462,6214,9359,70028,9455,5657,9278,9417,9380,9463,9468,9379,9419,9358,9415,70118,70013,6279,9465,9490,70122,9354,9428,9416,5650,9850,9055,70012,9383,9401,9831,9373,9405,9378,9845,9846,70172,9436,9059,9403,9384,9543,70173,9467,9469,9340,9345,9364,5130,9821,9427,9342,9343,9374,9473,9341,5656,9422,70171,9382,9389,70174,6427,9344,9402"; Version=1; Domain=co.uk; Max-Age=31536000; Expires=Fri, 26-Jun-2020 14:10:56 GMT; Path=/
Cache-Control: no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: Thu, 01 Dec 1994 16:00:00 GMT
Set-Cookie: dwsid=CrrDkasQl2bkQkNTIYvF2j8ne4drdWJWhADjhXBeiaTluEjaVL9wxCwS_xZwuCDcHtotCzEzGTGmZEQ-YjO19w==; path=/; HttpOnly
vary: accept-encoding
X-Content-Type-Options: nosniff
Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
Server: cloudflare
CF-RAY: 4ed801e00c87bfca-MAN
```

Note the `PM_unread_Onlineshop` cookie scope is `Domain=co.uk`.

i.e. upon receiving this cookie, the H3 cookie store is dutifully storing it under the `co.uk` domain. After which, every request to any `co.uk` domain will include that cookie.

This is _presumably_ inconsistent with how browsers handle cookies, and we need to make sure H3 knows about private v. public suffixes.

**Comment by anjackson** [comment]
From the console, these can be inspected with

```
cookies = appCtx.getBean("cookieStorage").hostSubset("co.uk");
rawOut.println(cookies);
```

And cleared with 

```
cookies = appCtx.getBean("cookieStorage").hostSubset("co.uk");
cookies.clear();
rawOut.println(cookies);
```

So I can do that manually for now.

**Comment by anjackson** [comment]
e.g.

```
suffs = ["ac.uk", "co.uk", "gov.uk", "ltd.uk", "me.uk", "net.uk", "nhs.uk", "org.uk", "plc.uk", "police.uk", "sch.uk" ]

suffs.each{ 
  cookies = appCtx.getBean("cookieStorage").hostSubset(it);
  rawOut.println(cookies);
  cookies.clear()
  rawOut.println(cookies);
}
```

**Comment by anjackson** [comment]
The Apache [BasicCookieStore](https://github.com/apache/httpcomponents-client/blob/4.3.x/httpclient/src/main/java/org/apache/http/impl/client/BasicCookieStore.java) looks much simpler, using this [CookieIdentityComparator](https://github.com/apache/httpcomponents-client/blob/37c80064726f0e4ae547c14ad73129483a81a67c/httpclient5/src/main/java/org/apache/hc/client5/http/cookie/CookieIdentityComparator.java).

The BdbCookieStore is [more complex](https://github.com/internetarchive/heritrix3/blob/b68731e2b9217e6d14ae2a5c283ebefb04543d6e/modules/src/main/java/org/archive/modules/fetcher/BdbCookieStore.java#L167-L192), iterating through parent domains.

**Comment by anjackson** [comment]
Okay, so @ato pointed out that cookie policy happens elsewhere, leading to this version note for HTTP Client 4.4:

> 4.4 "Default SSL hostname verifier and default cookie policy now validate certificate identity and cookie domain of origin against the public suffix list maintained by Mozilla.org <https://publicsuffix.org/list>"

So, this turns out to be basically the same as under #245 -- If we upgrade, this will improve things.

**Comment by anjackson** [comment]
Closed by #397.
